### PR TITLE
AMQPLIB: exchange types are limited

### PR DESCRIPTION
**Pull request by XVincentX**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

### CODE REVIEW

**Comment by danger-public** [comment]
Inspecting the JavaScript source for this package found some properties that are not in the .d.ts files.
The check for missing properties isn't always right, so take this list as advice, not a requirement.
## amqplib ([<kbd>unpkg</kbd>](https://unpkg.com/browse/amqplib@latest/))
was missing the following properties:
1. IllegalOperationError
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against a57903fac296ec19fe8ebab2556173f70fbe55e2
</p>

**Comment by XVincentX** [comment]
@mnahkies @abreits @nfantone @zelein @elibarzilay 

Can anybody checks this out?

**Comment by XVincentX** [comment]
@mnahkies @abreits @nfantone @nickzelei  @elibarzilay ping

**Comment by nickzelei** [comment]
This names seem to _mostly_ match what `amqplib` natively supports. These mostly look like Rabbit topic names though.

I was looking at the rabbit docs at what they detail for the Exchange Types and this list doesn't appear to be definitive.

See the table below, but this may be missing `'' | match`
I'm not 100% on this though.

https://www.rabbitmq.com/tutorials/amqp-concepts.html#exchanges

Exchange type | Default pre-declared names
-- | --
Direct exchange | (Empty string) and amq.direct
Fanout exchange | amq.fanout
Topic exchange | amq.topic
Headers exchange | amq.match (and amq.headers in RabbitMQ)

**Comment by nickzelei** [comment]
Also, looks like my github name needs updated in the comments. :sweat_smile:

File: types/amqplib/index.d.ts [hunk 1]
```diff
@@ -30,7 +34,7 @@ export interface Channel extends events.EventEmitter {
     bindQueue(queue: string, source: string, pattern: string, args?: any): Promise<Replies.Empty>;
     unbindQueue(queue: string, source: string, pattern: string, args?: any): Promise<Replies.Empty>;
 
-    assertExchange(exchange: string, type: string, options?: Options.AssertExchange): Promise<Replies.AssertExchange>;
+    assertExchange(exchange: string, type: 'direct' | 'topic' | 'headers' | 'fanout', options?: Options.AssertExchange): Promise<Replies.AssertExchange>;
```
Comment (nickzelei) [inline types/amqplib/index.d.ts:37]:
```suggestion
    assertExchange(exchange: string, type: '' | 'direct' | 'topic' | 'headers' | 'fanout' | 'match', options?: Options.AssertExchange): Promise<Replies.AssertExchange>;
```

**Review by nickzelei — CHANGES_REQUESTED** [review]

**Review by nickzelei — APPROVED** [review]

**Comment by yotamshapira** [comment]
@nickzelei , @XVincentX - what happens if I want to use a `x-delayed-message` type or any other type that comes from a plugin?

**Comment by nickzelei** [comment]
Is there a list somewhere that has a more inclusive list of types that we could update it to use? Otherwise we could append `| string` to widen the type while also providing hints for acceptable types.

**Comment by yotamshapira** [comment]
> Is there a list somewhere that has a more inclusive list of types that we could update it to use? Otherwise we could append `| string` to widen the type while also providing hints for acceptable types.

I think adding `| string` is the way to go since anyone can write a plugin that adds a new type: https://www.rabbitmq.com/plugins.html#basics (See first paragraph where it says "additional AMQP 0-9-1 exchange types")

**Comment by ticup** [comment]
Just a bump on the issue @yotamshapira raised.
We use the [delay-plugin](https://github.com/rabbitmq/rabbitmq-delayed-message-exchange), which uses the `x-delayed-message` type. Consequently, the last update of the types lib broke our build, so we had to downgrade again.

**Comment by nickzelei** [comment]
In progress on approving #49176

**Comment by ikonduktor** [comment]
So what about 'x-delayed-message'?
Now I have to uninstall amqplib because of this beautiful types; GJ

**Pull request merged**